### PR TITLE
Added support for providing replicas

### LINKED ISSUE #19: Allow specifying replica count for service
OpenCompose user should be allowed to specify number of replicas for the service to be able to scale it.

Proposal: Add an *optional* field named `replicas` into `ServiceSpec`. It defaults to value `1` and it should map into generated deployment under [DeploymentSpec.replicas](https://kubernetes.io/docs/api-reference/v1.5/#deploymentspec-v1beta1)

### PR DESCRIPTION
**Pull request by surajssd**
A user can now provide the number of replicas of the
service that are needed, using a field called `replicas`
in opencompose `service` section.

This is an optional field, if nothing is provided then
the value defaults to 1.

Fixes #19

### CODE REVIEW

**Comment by kadel** [comment]
Commented in the original issue https://github.com/redhat-developer/opencompose/issues/19#issuecomment-285638138

**Review by tnozicka — CHANGES_REQUESTED** [review]
- https://github.com/redhat-developer/opencompose/issues/19#issuecomment-285638138
- missing unit tests

**Comment by tnozicka** [comment]
@surajssd FYI and to save you time, you need to move from `int32` to `*int32`.

File: pkg/cmd/convert.go [hunk 1]
```diff
@@ -75,7 +75,7 @@ func GetValidatedObject(v *viper.Viper, cmd *cobra.Command, out, outerr io.Write
 
 		o, err := decoder.Decode(data)
 		if err != nil {
-			return nil, fmt.Errorf("could not unmarsha data for file '%s': %s", file, err)
+			return nil, fmt.Errorf("could not unmarshal data for file '%s': %s", file, err)
```
Comment (tnozicka) [inline pkg/cmd/convert.go]: Could you please make it a different commit as it has no relation to *Added support for providing replicas*?

Comment (surajssd) [reply pkg/cmd/convert.go]: done!

File: pkg/transform/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -106,6 +106,12 @@ func (t *Transformer) CreateDeployments(o *object.Service) ([]runtime.Object, er
 		},
 	}
 
+	// Set replica value only if it is given
+	if o.Replicas > 0 {
+		replicas := o.Replicas
+		d.Spec.Replicas = &replicas
```
Comment (surajssd) [inline pkg/transform/kubernetes/kubernetes.go]: This is a better thing to do or should I do: `d.Spec.Replicas = &o.Replicas` ?
Thoughts: @tnozicka @containscafeine ?

File: pkg/encoding/v1/encoding.go [hunk 3]
```diff
@@ -319,6 +320,9 @@ func (d *Decoder) Decode(data []byte) (*object.OpenCompose, error) {
 			Name: string(s.Name),
 		}
 
+		// user given value of replicas
+		os.Replicas = s.Replicas
```
Comment (surajssd) [inline pkg/encoding/v1/encoding.go]: using values as it is here and checking down whether it should land in `deployment` or not? thoughts: @tnozicka

File: pkg/encoding/v1/encoding_test.go [hunk 4]
```diff
@@ -397,6 +397,30 @@ volumes: []
...
+				Version: Version,
+				Services: []object.Service{
+					{
+						Name: "helloworld",
+						Containers: []object.Container{
+							{
+								Image: "tomaskral/nonroot-nginx",
+							},
+						},
+						Replicas: 0,
```
Comment (surajssd) [inline pkg/encoding/v1/encoding_test.go]: since using values as it is so not much changes, checks done at a level up when creating `deployment`, so original info is not lost from opencompose as well as in cluster.

File: pkg/transform/kubernetes/kubernetes.go (hunk 2 above)
Comment (concaf) [reply pkg/transform/kubernetes/kubernetes.go]: @surajssd `d.Spec.Replicas = &o.Replicas` should be fine, no?

Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes.go]: @containscafeine i thought it will have to save entire object `o` to just assign value, until `d` is alive. But if I can copy it in some variable then it will be less wastage of resources?

Comment (tnozicka) [reply pkg/transform/kubernetes/kubernetes.go]: @surajssd `d.Spec.Replicas = o.Replicas` you should make o.Replicas pointer as well because it will be a pointer from the parser too

Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes.go]: oh yes, makes sense

File: pkg/encoding/v1/encoding.go [hunk 5]
```diff
@@ -184,6 +184,7 @@ func (c *Container) UnmarshalYAML(unmarshal func(interface{}) error) error {
 type Service struct {
 	Name       ResourceName `yaml:"name"`
 	Containers []Container  `yaml:"containers"`
+	Replicas   int32        `yaml:"replicas,omitempty"`
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding.go]: it has to be a pointer here otherwise you can't distinguish between user setting 0 and not specifying the value

**Review by tnozicka — CHANGES_REQUESTED** [review]

File: pkg/encoding/v1/encoding.go (hunk 5 above)
Comment (surajssd) [reply pkg/encoding/v1/encoding.go]: done!

File: pkg/encoding/v1/encoding.go (hunk 3 above)
Comment (tnozicka) [reply pkg/encoding/v1/encoding.go]: probably out-of date now

File: pkg/transform/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -106,6 +106,11 @@ func (t *Transformer) CreateDeployments(o *object.Service) ([]runtime.Object, er
 		},
 	}
 
+	// Set replica value only if it is given
+	if o.Replicas != nil {
+		d.Spec.Replicas = o.Replicas
```
Comment (tnozicka) [inline pkg/transform/kubernetes/kubernetes.go]: only L111 is needed since it's now a pointer

File: pkg/encoding/v1/encoding_test.go [hunk 7]
```diff
@@ -419,5 +419,44 @@ volumes: []
 			continue
 		}
 	}
+}
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding_test.go]: I don't like these unit test. We do table driven testing (a bit of Golang way) everywhere else.

There should be 2 unit tests for:
 - all the test cases (nil, 0, 42, fail, ...) - TestService_UnmarshalYAML (you can get inspiration [here](https://github.com/tnozicka/opencompose/blob/a21430158830bff240a8252aa4b609e994907238/pkg/encoding/v1/encoding_test.go#L118))
The unit test for service hasn't been there before which is my fault, but If you wouldn't mind adding it there, that will be great 
 - to ensure the value gets copied (probably just add one case to existing test) -https://github.com/surajssd/opencompose/blob/8e34ab28588e02b85fada5cc52661cf300f36b0a/pkg/encoding/v1/encoding_test.go#L147

File: pkg/encoding/v1/encoding.go [hunk 8]
```diff
@@ -319,6 +320,9 @@ func (d *Decoder) Decode(data []byte) (*object.OpenCompose, error) {
 			Name: string(s.Name),
 		}
 
+		// user given value of replicas
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding.go]: it's now a pointer to the value; maybe just drop the comment since its obvious anyway

**Review by tnozicka — CHANGES_REQUESTED** [review]

File: pkg/encoding/v1/encoding_test.go [hunk 9]
```diff
@@ -397,6 +465,54 @@ volumes: []
...
+					},
+				},
+			},
+		},
+		{
+			true, `
+version: 0.1-dev
+services:
+- name: helloworld
+  replicas: -1
```
Comment (surajssd) [inline pkg/encoding/v1/encoding_test.go]: even after giving value as negative this test passes, since the check on negativity happens later? Apart from adding tests about createdeployment and verifying this actually fails @tnozicka do you have any thought on handling this well?

File: pkg/transform/kubernetes/kubernetes.go [hunk 10]
```diff
@@ -106,6 +106,16 @@ func (t *Transformer) CreateDeployments(o *object.Service) ([]runtime.Object, er
...
 
+	// Set replica value only if it is given
+	if o.Replicas != nil {
+		// and is positive number
+		if *o.Replicas >= 0 {
+			d.Spec.Replicas = o.Replicas
+		} else {
+			return nil, fmt.Errorf("Replica count is negative in service: %q", o.Name)
+		}
+	}
```
Comment (surajssd) [inline pkg/transform/kubernetes/kubernetes.go]: this is where I have added the check and this is needed otherwise the artifact is generated with negative number and k8s won't take that!

Comment (tnozicka) [reply pkg/transform/kubernetes/kubernetes.go]: actually, parser should be the one checking that since it's invalid input

File: pkg/encoding/v1/encoding_test.go [hunk 11]
```diff
@@ -144,9 +144,77 @@ func TestEnvVariable_UnmarshalYAML(t *testing.T) {
...
+replicas: notint
+containers:
+- image: tomaskral/kompose-demo-frontend:test
+`,
+			nil,
+		},
+	}
+
+	for _, test := range tests {
+		t.Logf("Running test: %q", test.Name)
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding_test.go]: please use t.Run where you can pass the name

we are moving to using subtests (https://blog.golang.org/subtests)

https://github.com/tnozicka/opencompose/blob/a21430158830bff240a8252aa4b609e994907238/pkg/encoding/v1/encoding_test.go#L207

File: pkg/encoding/v1/encoding_test.go (hunk 9 above)
Comment (tnozicka) [reply pkg/encoding/v1/encoding_test.go]: parser should check this

File: pkg/encoding/v1/encoding_test.go [hunk 12]
```diff
@@ -144,9 +144,77 @@ func TestEnvVariable_UnmarshalYAML(t *testing.T) {
...
+					},
+				},
+			},
+		},
+
+		{
+			"Giving replica as string to see parsing fails",
+			false, `
+name: frontend
+replicas: notint
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding_test.go]: probably a typo: s/notint/nonint/ or s/notint/non-int/

**Review by tnozicka — CHANGES_REQUESTED** [review]

File: pkg/encoding/v1/encoding_test.go (hunk 12 above)
Comment (surajssd) [reply pkg/encoding/v1/encoding_test.go]: It's just for showing this is 'not an integer' value

File: pkg/encoding/v1/encoding_test.go (hunk 7 above)
Comment (surajssd) [reply pkg/encoding/v1/encoding_test.go]: done!

File: pkg/transform/kubernetes/kubernetes.go (hunk 10 above)
Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes.go]: Added checks to parser as well

Comment (kadel) [reply pkg/transform/kubernetes/kubernetes.go]: If this is checked in parser, does it make sense to check it here again?

**Comment by kadel** [comment]
Can you please update docs (file-reference) as part of this PR?

**Comment by surajssd** [comment]
@kadel done!

File: docs/file-reference.md [hunk 13]
```diff
@@ -54,6 +56,14 @@ Each service has name and list of the containers.
 
 Name of the service.
 
+#### replicas
+| type  | required |
+|-------|----------|
+| int32 |    no    |
+
+Number of desired pods of this particluar service. This is an optional field.
+If nothing is provided the artifact generated won't have anything either.
```
Comment (kadel) [inline docs/file-reference.md]: I don't know about this sentence :-(
It won't say anything useful for people that are not familiar with Kubernetes.
I think that we should avoid mentioning generated artifacts as much as possible.

File: pkg/transform/kubernetes/kubernetes.go (hunk 10 above)
Comment (tnozicka) [reply pkg/transform/kubernetes/kubernetes.go]: @kadel agreed. That's what my previous comment was suppose to mean (move, not copy the check) - this isn't task transformer should do

File: docs/file-reference.md [hunk 14]
```diff
@@ -38,11 +39,12 @@ Version of OpenCompose format.
 Describes one service. Service can be composed of one or multiple containers which are scheduled 
 together and can communicate using localhost. (Containers in the same service share network namespace.)
 
-Each service has name and list of the containers.
+Each service has name, replicas and list of the containers.
```
Comment (tnozicka) [inline docs/file-reference.md]: it doesn't have to have `replicas`

File: docs/file-reference.md (hunk 13 above)
Comment (kadel) [reply docs/file-reference.md]: Can we just delete that last sentence sentence?

Comment (surajssd) [reply docs/file-reference.md]: done!

File: docs/file-reference.md (hunk 14 above)
Comment (surajssd) [reply docs/file-reference.md]: done!

File: pkg/transform/kubernetes/kubernetes.go (hunk 10 above)
Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes.go]: If I remove check from here, how do we address the condition where someone wants to use the opencompose as lib?

Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes.go]: and we wanna do the validation? ideally for an app like this where data is coming from multiple places we should have a layer which does it for us? WDYT @tnozicka ?

Comment (kadel) [reply pkg/transform/kubernetes/kubernetes.go]: May be we could use [object.Validate](https://github.com/redhat-developer/opencompose/blob/51b38c48fc8af3bb815e842629b700bfff5c6dd6/pkg/object/object.go#L50) for this.

All validation logic should be in one place

Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes.go]: this makes sense to me!

**Comment by surajssd** [comment]
@kadel @tnozicka updated as you guys mentioned, will squash when you guys give LGTM!

File: pkg/object/object.go [hunk 15]
```diff
@@ -45,9 +48,34 @@ type OpenCompose struct {
...
+	return nil
+}
+
+func (s *Service) Validate() error {
+	// validate service name, like it cannot have underscores, etc.
+
+	// validate containers
+
+	// validate replicas
+	if err := validateReplicas(s.Replicas); err != nil {
```
Comment (kadel) [inline pkg/object/object.go]: Stupid question as usual 😄 
If replica check is just one line, and it is used in one place, does it make sense to have a separate function for it? Wouldn't be code more readable if actual check was here? 😇

Comment (surajssd) [reply pkg/object/object.go]: I wanted to make it testable so separated it out!

Comment (kadel) [reply pkg/object/object.go]: than you testing separately one `if` statement? 
I don't know, but for me it looks like it is just overcomplicating things

`Validate` can be tested as whole.

Comment (surajssd) [reply pkg/object/object.go]: Even I thought this might be overdoing but then there arent other tests so just for testing replicas I will need to create entire stucts object.

And here I can do specific tests like these one: https://github.com/redhat-developer/opencompose/pull/66/files#diff-4f9edbaf8553598b5a269f232bab233cR16

File: pkg/transform/kubernetes/kubernetes_test.go [hunk 16]
```diff
@@ -344,3 +348,158 @@ func TestTransformer_CreateServices(t *testing.T) {
...
+					},
+				},
+			},
+		},
+		{
+			"When valid replica value not given, this matches because only valid data should come",
+			true,
+			&object.Service{
+				Name:     name,
+				Replicas: goutil.Int32Addr(-1),
```
Comment (surajssd) [inline pkg/transform/kubernetes/kubernetes_test.go]: @kadel also comment on this test?

File: pkg/encoding/v1/encoding_test.go [hunk 17]
```diff
@@ -144,9 +147,109 @@ func TestEnvVariable_UnmarshalYAML(t *testing.T) {
...
+			"Replica as negative int, should pass because this is just unmarshalling.",
+			true, `
+name: frontend
+replicas: -1
+containers:
+- image: tomaskral/kompose-demo-frontend:test
+`,
+			&Service{
+				Name:     "frontend",
+				Replicas: goutil.Int32Addr(-1),
```
Comment (surajssd) [inline pkg/encoding/v1/encoding_test.go]: @kadel also this test, please comment if this is needed ?

File: pkg/object/object.go (hunk 15 above)
Comment (kadel) [reply pkg/object/object.go]: As soon as you add another validation test to this you will have to test whole `Validate` function anyway.

Comment (surajssd) [reply pkg/object/object.go]: updated as per comments!

**Review by kadel — APPROVED** [review]
lgtm

**Comment by surajssd** [comment]
@kadel is it okay to merge?

**Comment by kadel** [comment]
> @kadel is it okay to merge?

lets wait for @tnozicka review

**Comment by surajssd** [comment]
ping @tnozicka

File: docs/file-reference.md [hunk 18]
```diff
@@ -54,6 +56,13 @@ Each service has name and list of the containers.
 
 Name of the service.
 
+#### replicas
+| type  | required |
+|-------|----------|
+| int32 |    no    |
```
Comment (tnozicka) [inline docs/file-reference.md]: wouldn't **u**int32 be more descriptive since any negative value is invalid. Or it would be good to mention it somewhere else

File: pkg/encoding/v1/encoding_test.go [hunk 19]
```diff
@@ -144,9 +147,90 @@ func TestEnvVariable_UnmarshalYAML(t *testing.T) {
...
+					{
+						Image: "tomaskral/kompose-demo-frontend:test",
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("Running test: %q", test.Name), func(t *testing.T) {
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding_test.go]: There should be just a test name; go test tool will make the appropriate announcement on what is running

File: pkg/encoding/v1/encoding_test.go [hunk 20]
```diff
@@ -144,9 +147,90 @@ func TestEnvVariable_UnmarshalYAML(t *testing.T) {
 	}
 }
 
+func TestService_UnmarshalYAML(t *testing.T) {
+
+	tests := []struct {
+		Name           string
+		ExpectedSucess bool
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding_test.go]: consistent naming with the rest would be `Succeed`

File: pkg/encoding/v1/encoding_test.go [hunk 21]
```diff
@@ -144,9 +147,90 @@ func TestEnvVariable_UnmarshalYAML(t *testing.T) {
...
+			err := yaml.Unmarshal([]byte(test.RawService), &service)
+			if err != nil {
+				if test.ExpectedSucess {
+					t.Errorf("Failed to unmarshal: %#v\nError: %#v", test.RawService, err)
+				}
+				return
+			}
+
+			if !test.ExpectedSucess {
+				t.Errorf("Expected %#v to fail, but succeeded!", test.RawService)
```
Comment (tnozicka) [inline pkg/encoding/v1/encoding_test.go]: you can merge these 2 lines by using t.Fatalf

File: pkg/goutil/goutil_test.go [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+		input  int32
+		output *int32
+	}{
+		{1, Int32Addr(1)},
+		{2, Int32Addr(2)},
+	}
+
+	for _, test := range tests {
+		if test.input != *test.output {
+			t.Errorf("The input and output doesn't match, for %d", test.input)
```
Comment (tnozicka) [inline pkg/goutil/goutil_test.go]: logically you should print both of those values to know what's wrong, not just input

File: pkg/object/object.go [hunk 23]
```diff
@@ -45,9 +48,27 @@ type OpenCompose struct {
 	Volumes  []Volume
 }
 
+func (s *Service) Validate() error {
+	// validate service name, like it cannot have underscores, etc.
+
+	// validate containers
+
+	// validate replicas
+	if s.Replicas != nil && *s.Replicas < 0 {
```
Comment (tnozicka) [inline pkg/object/object.go:59]: such checks should be done in the parser as I have mentioned in my previous comments. it's n invalid input for the parser and it is the only place we can have more info like a line number - and all the other checks are done in parser as well!

File: pkg/object/object.go [hunk 24]
```diff
@@ -45,9 +48,27 @@ type OpenCompose struct {
...
+		return fmt.Errorf("Replica count is negative in service: %q", s.Name)
```
Comment (tnozicka) [inline pkg/object/object.go:60]: s/is/shall not be/

File: pkg/transform/kubernetes/kubernetes_test.go [hunk 25]
```diff
@@ -344,3 +348,121 @@ func TestTransformer_CreateServices(t *testing.T) {
...
+			"service": name,
+		},
+	}
+	strategy := ext_v1beta1.DeploymentStrategy{
+		Type: ext_v1beta1.RollingUpdateDeploymentStrategyType,
+	}
+
+	tests := []struct {
+		Name           string
+		ExpectSuccess  bool
```
Comment (tnozicka) [inline pkg/transform/kubernetes/kubernetes_test.go]: non-consistent naming as above

File: pkg/transform/kubernetes/kubernetes_test.go [hunk 26]
```diff
@@ -344,3 +348,121 @@ func TestTransformer_CreateServices(t *testing.T) {
...
+						},
+					},
+				},
+			},
+		},
+	}
+
+	transformer := Transformer{}
+	for _, test := range tests {
+
```
Comment (tnozicka) [inline pkg/transform/kubernetes/kubernetes_test.go]: s/\n//

File: pkg/transform/kubernetes/kubernetes_test.go [hunk 27]
```diff
@@ -344,3 +348,121 @@ func TestTransformer_CreateServices(t *testing.T) {
...
+		t.Run(fmt.Sprintf("Running test: %q", test.Name), func(t *testing.T) {
```
Comment (tnozicka) [inline pkg/transform/kubernetes/kubernetes_test.go]: there should be just name of the test

**Review by tnozicka — DISMISSED** [review]

File: pkg/transform/kubernetes/kubernetes_test.go (hunk 26 above)
Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes_test.go]: done!

File: pkg/transform/kubernetes/kubernetes_test.go (hunk 27 above)
Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes_test.go]: done!

File: pkg/transform/kubernetes/kubernetes_test.go (hunk 25 above)
Comment (surajssd) [reply pkg/transform/kubernetes/kubernetes_test.go]: done!

File: pkg/goutil/goutil_test.go (hunk 22 above)
Comment (surajssd) [reply pkg/goutil/goutil_test.go]: done

File: pkg/encoding/v1/encoding_test.go (hunk 20 above)
Comment (surajssd) [reply pkg/encoding/v1/encoding_test.go]: done

File: pkg/encoding/v1/encoding_test.go (hunk 19 above)
Comment (surajssd) [reply pkg/encoding/v1/encoding_test.go]: done

File: pkg/encoding/v1/encoding_test.go (hunk 21 above)
Comment (surajssd) [reply pkg/encoding/v1/encoding_test.go]: done

File: docs/file-reference.md (hunk 18 above)
Comment (surajssd) [reply docs/file-reference.md]: should the internal object be also changed to `uint32`?

File: pkg/object/object.go (hunk 23 above)
Comment (surajssd) [reply pkg/object/object.go:59]: Logically, the value is valid for parser because parser is only supposed to parse data for its type and not validate if it is correct or not. That's the job of validator.

And I would like to have validation happening in only one place not twice. Someone using opencompose as lib might not be passing validated data so how are we gonna validate that?

File: docs/file-reference.md (hunk 18 above)
Comment (kadel) [reply docs/file-reference.md]: I think `int32` in the code is ok.

I'm just thinking if it makes sense in documentation mention `int32` type, maybe just `integer`?

File: pkg/object/object.go (hunk 23 above)
Comment (kadel) [reply pkg/object/object.go:59]: Letes move on, and go with this as it is.
We can figure out how we are going to do all validations in another PR.

https://github.com/redhat-developer/opencompose/issues/92

Comment (surajssd) [reply pkg/object/object.go:59]: So we can merge this one?

Comment (kadel) [reply pkg/object/object.go:59]: see my other comment regarding docs (https://github.com/redhat-developer/opencompose/pull/66/files#r110106733)
than if you squash it, i think we can merge it

Comment (surajssd) [reply pkg/object/object.go:59]: Sure on it now!

File: docs/file-reference.md (hunk 18 above)
Comment (surajssd) [reply docs/file-reference.md]: done

**Review by kadel — APPROVED** [review]
